File: algorithms/N/Number of Valid Words for Each Puzzle/Number of Valid Words for Each Puzzle.py
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        look_up=collections.defaultdict(int)
        def get_mask(word):
            mask=0
            for c in word:
                mask |= 1<<(ord(c)-ord('a'))
            return mask
        for word in words:
            mask=get_mask(word)
            look_up[mask]+=1
        ans=[]
        def solve(puzzle_idx,mask,c_idx):
            if c_idx==len(puzzles[puzzle_idx]):
                ans[-1]+=look_up[mask]
                return
            #take this c_idx
            solve(puzzle_idx,mask | 1<<(ord(puzzles[puzzle_idx][c_idx])-ord('a')),c_idx+1)
            #dont take this c_idx
            solve(puzzle_idx,mask,c_idx+1)
        for i,puzzle in enumerate(puzzles):
            ans.append(0)
            solve(i,1<<(ord(puzzle[0])-ord('a')),1)
        return ans

File: scripts/algorithms/N/Number of Valid Words for Each Puzzle/Number of Valid Words for Each Puzzle.cpp
// Runtime: 392 ms (Top 85.55%) | Memory: 69 MB (Top 97.11%)
class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
        unordered_map<int, int> maskCnt;
        for(const auto& w: words) ++maskCnt[bitmask(w)];
        vector<int> ans(puzzles.size(), 0);
        for(int i = 0; i < puzzles.size(); i++){
            int mask, subMask = mask = bitmask(puzzles[i]), first = bitmask(puzzles[i].substr(0,1));
            do{
                if( (first & subMask) == first && maskCnt.count(subMask)) ans[i] += maskCnt[subMask]; //ok
            }while(subMask = (subMask - 1) & mask);
        }
        return ans;
    }

private:
    int bitmask(const string& word, int mask = 0){
        for(auto c: word) mask |= (1 << (c - 'a'));
        return mask;
    }
};